rfcs/0009_Unified_Static_Memory_Planning.md

```diff
@@ -515,4 +663,6 @@ NOTE : to support tir.constants generally, we'll be enhancing the bound relay.co
 
 # Drawbacks
 
-* The relay "main" function that describes the call order to operator PrimFuncs has to be described in TIR to be able to integrate the USMP into the respective executor codegen. However, we dont view this as a major problem as the relay "main" function could easily be lowered to TIR.
\ No newline at end of file
+* The relay "main" function that describes the call order to operator PrimFuncs has to be described in TIR to be able to integrate the USMP into the respective executor codegen. However, we dont view this as a major problem as the relay "main" function could easily be lowered to TIR.
+
+* The U4 usecase will only be supported with [Embedded C Runtime Interface](https://discuss.tvm.apache.org/t/rfc-utvm-embedded-c-runtime-interface/9951/14). This is mainly because the nature of the requirement is associated with embedded usecases. However, the USMP changes here should be complimentary to support other runtime interfaces such as Module-based Model Runtime Interface's set_input and set_output in future.
```

Comment:
if we did implement this for e.g. C++ runtime or a scenario where `DLTensor` were passed (perhaps we have dynamic shapes), would we allocate DLTensor instances in whatever section of workspace we reserve for them? conceptually we could just consider those `DLTensor`s as companion buffers which need to be live at the same time as the `data`.

Reply:
Good point again!

This is generally applicable for Embedded C Runtime Inferface as well.

Another way to frame the the suggestion is to say : what cant we treat the buffer space used by I/O tensors as seperate workspace buffers ? 

I have considered this too :).  The short anwser is it is sub-optimal use disjoint workspace buffers rather than a single workspace buffer -- because it does not allow partial overlaps of tensors across such buffers. TLDR; 

So the workspace buffer is viewed as a spatio-temporal buffer and tensors are considered as spatio-temporal allocation. Having disjoint workspace buffers, create a spatial barrier that could potentially lead to higher fragmentation. Therefore, out of the two options; we are prioritizing the better one here. Again, this could be an another usecase if there is value for it.

Again, I can add few lines for this as future consideration.

Reply:
i agree it's probably sub-optimal to separate I/O buffers from intermediates, but what I mean here is: how would we extend the planner to also allocate memory for the `DLTensor` metadata inside the workspace buffer? the Executor currently has to allocate memory for those somewhere in the case where we need that metadata (e.g. --use-unpacked-api=false or we have dynamic shapes).

Reply:
@areusch smart pointer (DLTensor) metadata is not something considered so far.

I admit that we need to account for a certain overhead for each buffers based on the calling convention used. Currently the USMP only focus are about planning for .data of DLTensors -- however, I feel it is orthogonal to the proposal here and applicable generally to all usecases.

Reply:
> I admit that we need to account for a certain overhead 

i guess the main question is: do you think that DLTensor data should be "overhead" in the sense that USMP should consider a buffer to need `sizeof(data) + sizeof(DLTensor)` bytes contiguously, or do you prefer to keep with all the `data` in one contiguous block and the DLTensors elsewhere? the second option here seems more inline with the unpacked calling convention.